### PR TITLE
add track info for mem tracker fault inject

### LINKED ISSUE #6158: add track info for mem tracker fault inject
## Enhancement
add track info for mem tracker fault inject

### PR DESCRIPTION
**Pull request by bestwoody**
### What problem does this PR solve?

Issue Number: close #6158

Problem Summary:

### What is changed and how it works?
add track info for mem tracker fault inject, such as num of threads, peak mem usage not tracked by Processlist.
simplify logic of real rss check
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by yibin87** [comment]
File an issue?

File: dbms/src/Common/MemoryTracker.cpp [hunk 1]
```diff
@@ -207,6 +210,8 @@ __thread MemoryTracker * current_memory_tracker = nullptr;
 thread_local MemoryTracker * current_memory_tracker = nullptr;
 #endif
 
+std::shared_ptr<MemoryTracker> root_of_mem_trackers_not_managed_by_process_list = MemoryTracker::create();
```
Comment (yibin87) [inline dbms/src/Common/MemoryTracker.cpp:213]: change a name like root_of_storage_mem_trackers to make it more specfic?  Just my personal opinion.

File: dbms/src/Common/MemoryTracker.cpp [hunk 2]
```diff
@@ -22,7 +22,7 @@
 
 #include <iomanip>
 
-std::atomic<Int64> real_rss{0};
+std::atomic<Int64> real_rss{0}, proc_num_threads{1};
```
Comment (yibin87) [inline dbms/src/Common/MemoryTracker.cpp:25]: why proc_num_threads default value is 1?

File: dbms/src/Storages/BackgroundProcessingPool.cpp [hunk 3]
```diff
@@ -155,6 +155,7 @@ void BackgroundProcessingPool::threadFunction(size_t thread_idx)
     }
 
     auto memory_tracker = MemoryTracker::create();
+    memory_tracker->setNext(root_of_mem_trackers_not_managed_by_process_list.get());
```
Comment (windtalker) [inline dbms/src/Storages/BackgroundProcessingPool.cpp:158]: Looks like `root_of_mem_trackers_not_managed_by_process_list` is only used by `BackgroundProcessingPool`, why not just define it as a static member in `BackgroundProcessingPool`?

Comment (bestwoody) [reply dbms/src/Storages/BackgroundProcessingPool.cpp:158]: for further usage，we need a “other” mem tracker to control those not managed by ProcList

Comment (windtalker) [reply dbms/src/Storages/BackgroundProcessingPool.cpp:158]: OK, maybe we can rename it to `root_of_non_query_mem_trackers`?

Comment (bestwoody) [reply dbms/src/Storages/BackgroundProcessingPool.cpp:158]: good idea

File: dbms/src/Common/MemoryTracker.cpp (hunk 1 above)
Comment (bestwoody) [reply dbms/src/Common/MemoryTracker.cpp:213]: ok

Comment (bestwoody) [reply dbms/src/Common/MemoryTracker.cpp]: done

File: dbms/src/Storages/BackgroundProcessingPool.cpp (hunk 3 above)
Comment (bestwoody) [reply dbms/src/Storages/BackgroundProcessingPool.cpp]: done

**Comment by bestwoody** [comment]
> File an issue?

added

**Review by windtalker — APPROVED** [review]
LGTM

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Common/BackgroundTask.cpp [hunk 4]
```diff
@@ -74,13 +73,17 @@ void CollectProcInfoBackgroundTask::begin()
...
-        process_mem_usage(resident_set);
+        process_mem_usage(resident_set, cur_proc_num_threads, cur_virt_size);
         resident_set *= 1024; // unit: byte
         real_rss = static_cast<Int64>(resident_set);
-
-        usleep(100000); // sleep 100ms
+        proc_num_threads = cur_proc_num_threads;
+        proc_virt_size = cur_virt_size;
+        baseline_of_query_mem_tracker = root_of_query_mem_trackers->get();
+        usleep(10000); // sleep 10ms
```
Comment (windtalker) [inline dbms/src/Common/BackgroundTask.cpp:86]: Why change this?

File: dbms/src/Common/BackgroundTask.cpp [hunk 5]
```diff
@@ -74,13 +73,17 @@ void CollectProcInfoBackgroundTask::begin()
...
     {
-        process_mem_usage(resident_set);
+        process_mem_usage(resident_set, cur_proc_num_threads, cur_virt_size);
         resident_set *= 1024; // unit: byte
         real_rss = static_cast<Int64>(resident_set);
-
-        usleep(100000); // sleep 100ms
+        proc_num_threads = cur_proc_num_threads;
+        proc_virt_size = cur_virt_size;
+        baseline_of_query_mem_tracker = root_of_query_mem_trackers->get();
```
Comment (windtalker) [inline dbms/src/Common/BackgroundTask.cpp:85]: Assign a pointer to Int64?

File: dbms/src/Common/MemoryTracker.cpp [hunk 6]
```diff
@@ -111,7 +116,7 @@ void MemoryTracker::alloc(Int64 size, bool check_memory_limit)
         Int64 current_bytes_rss_larger_than_limit = bytes_rss_larger_than_limit.load(std::memory_order_relaxed);
         bool is_rss_too_large = (!next.load(std::memory_order_relaxed) && current_limit
                                  && real_rss > current_limit + current_bytes_rss_larger_than_limit
-                                 && will_be > current_limit - (real_rss - current_limit - current_bytes_rss_larger_than_limit));
+                                 && will_be > baseline_of_query_mem_tracker);
```
Comment (windtalker) [inline dbms/src/Common/MemoryTracker.cpp:119]: compare with address?

File: dbms/src/Common/BackgroundTask.cpp (hunk 4 above)
Comment (bestwoody) [reply dbms/src/Common/BackgroundTask.cpp:86]: the old interval is too large. 10ms can get a latest metrics of proc

File: dbms/src/Common/BackgroundTask.cpp (hunk 5 above)
Comment (bestwoody) [reply dbms/src/Common/BackgroundTask.cpp:85]: it's `root_of_query_mem_trackers->get() ` not `root_of_query_mem_trackers.get()`. `root_of_query_mem_trackers->get()` get the amount of mem tracker.

File: dbms/src/Common/MemoryTracker.cpp (hunk 6 above)
Comment (bestwoody) [reply dbms/src/Common/MemoryTracker.cpp:119]: type of `baseline_of_query_mem_tracker`  is `atomic<int64>`

**Pull request merged**